### ISSUE TITLE
ENV loses some special characters (e.g., $) after commit

### ISSUE THREAD

**Issue by EricHripko** [containers/buildah#1556]
**Description**

`buildah` loses some special characters (namely, `$`) in values of environment variables defined with `ENV`. This only happens after the commit - the values are fine during build.

**Steps to reproduce the issue:**
1. Use the following `Dockerfile`
```Dockerfile
FROM docker.io/ubuntu
ENV LIB="$(PREFIX)/lib"
RUN env
```
2. Build an image: `buildah bud -t proof2 .`
3. Create a container from it: `buildah from proof2`
4. Inspect environment variables: `buildah run proof2-working-container env`

**Describe the results you received:**
The following output. Note that `LIB` environment variable lost `$` and is now `(PREFIX)/lib` instead of `$(PREFIX)/lib`.
```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LIB=(PREFIX)/lib
HOSTNAME=6e9977f521b5
HOME=/root
```

**Describe the results you expected:**
The following output:
```
HOSTNAME=14d7241306c4
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LIB=$(PREFIX)/lib
PWD=/
```

**Output of `rpm -q buildah` or `apt list buildah`:**

```
Listing... Done
buildah/xenial,now 1.8-1~dev~ubuntu16.04.2~ppa5 amd64 [installed]
```

**Output of `buildah version`:**

```
Version:         1.8-dev
Go Version:      go1.10.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:
Git Commit:
Built:           Sat Apr 13 16:42:38 2019
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
N/A
```

**Output of `cat /etc/*release`:**

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
```

**Output of `uname -a`:**

```
Linux scw-35eb3f 4.5.7-docker-4 #1 SMP Tue Jul 12 09:41:14 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information

# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary read-write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# AdditionalImageStores is used to pass paths to additional read-only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers (currently overlay, zfs, vfs, btrfs)
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"
```

**Comment by TomSweeneyRedHat** [comment]
@EricHripko Thanks for the heads up.  Do you by chance know if this occurred in earlier versions of Buildah?  I ask as we changed our CLI handling software as of v1.7 and this smells like it might be related.  If you don't know, no worries, we'll take a look regardless.

**Comment by EricHripko** [comment]
Happy to help! I'll try to troubleshoot this further if I have some time next week.

Unfortunately, I am not aware if this happened in earlier versions of `buildah` since I've not used anything but `1.8-dev`.

**Comment by rhatdan** [comment]
It seems to be working for me with buildah-1.8.0 in Fedora
```
buildah run  56a2c8de44a19a166686587a37072d424d5bec792a1f671f0d5782302e46712b-working-container env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LIB=$(PREFIX)/lib
HOSTNAME=1394bd9feeb4
HOME=/root
```

**Comment by EricHripko** [comment]
I can reliably reproduce the issue on the machine reported above (xenial):
```
$ buildah run proof2-working-container env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LIB=(PREFIX)/lib
HOSTNAME=03ae106f7236
HOME=/root
```

I've built the tip of master on a different machine (disco) and it works correctly now:
```
$ ./buildah run proof2-working-container env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LIB=$(PREFIX)/lib
HOSTNAME=36f0b7d38a18
HOME=/root
```
Unfortunately, there doesn't seem to be an official build of `buildah` in projectatomic PPA for disco - so I am unsure whether this is because of a different OS version or if some recent changes fixed this issue.

**Comment by rhatdan** [comment]
This means that it is fixed in buildah-1.8.0, which was just released yesterday.  We will get a new version out for Ubuntu shortly

**Comment by EricHripko** [comment]
I had a closer look. It looks like the culprit surprisingly is the version of go. On `e9184eaac6615cd3c20ed5bbdc44e12c2d5f5f56` commit:
- I can reproduce the issue with go1.10.4 (same version as in the issue reported)
- I cannot reproduce the issue with go1.12.4 (version I used for dev builds)

**Comment by rhatdan** [comment]
So is it fixed in the latest versions we have built for Ubuntu?

**Comment by EricHripko** [comment]
I am already on the latest version available (it's the same one as the issue was reported on). I don't see any newer versions in the [Project Atomic PPA](https://launchpad.net/~projectatomic/+archive/ubuntu/ppa?field.series_filter=xenial). 
```
$ apt-get update && apt-get upgrade buildah
...
buildah is already the newest version (1.8-1~dev~ubuntu16.04.2~ppa5).
...
$ buildah version
Version:         1.8-dev
Go Version:      go1.10.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  
Git Commit:      
Built:           Sat Apr 13 16:42:38 2019
OS/Arch:         linux/amd64
$ buildah bud -t proof .
STEP 1: FROM docker.io/ubuntu
...
Storing signatures
--> f3e6b9f4609167516338318e995c152d01391db36294e1bdbb567959f44a518d
$ buildah inspect proof | grep PREFIX
...
                "LIB=(PREFIX)/lib"
...
                "LIB=(PREFIX)/lib"
                "LIB=(PREFIX)/lib"
...
```

**Comment by rhatdan** [comment]
Well we should be at 1.8.2 for this and hopefully that is built with newer versions of the go compiler.

@lsm5 @jnovy PTAL

**Comment by EricHripko** [comment]
I reproduced this with the following small program:
```go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	k := "LIB"
	v := "$(PREFIX)"
	// config.go:271
	reset := func(s *[]string) {
		getenv := func(name string) string {
			for i := range *s {
				val := strings.SplitN((*s)[i], "=", 2)
				if len(val) == 2 && val[0] == name {
					return val[1]
				}
			}
			return name
		}
		n := []string{}
		for i := range *s {
			if !strings.HasPrefix((*s)[i], k+"=") {
				n = append(n, (*s)[i])
			}
			v = os.Expand(v, getenv)
		}
		n = append(n, k+"="+v)
		*s = n
	}
	// config.go:290

	m := []string{"PATH=/usr/bin"}
	reset(&m)
	fmt.Print(m)
}
```

```
$ go1.10.4 run test.go
[PATH=/usr/bin LIB=(PREFIX)]
$ go run test.go
[PATH=/usr/bin LIB=$(PREFIX)]⏎ 
```

This is a behaviour change in [go1.11](https://github.com/golang/go/commit/7bf631e1fc55839a6bff404a91597326e1addc10#diff-b30ffd1cf7612ccebac97fced83d5106) described in https://github.com/golang/go/issues/24345. 

My suggestion on how to address this:
- Specify that version 1.11 or higher is required here: https://github.com/containers/buildah/blob/master/install.md#installation-from-github
- Add a test to ensure correct behaviour

Does that sound fine? If so, I'll put in a PR.

**Comment by rhatdan** [comment]
SGTM

**Comment by EricHripko** [comment]
Looks like after #1607, version restrictions are no longer necessary.
I've put a PR for adding a test above, as discussed previously.

**Comment by EricHripko** [comment]
Resolved in #1608, closing.
